p4-farm-nft/src/lib.rs

```diff
@@ -0,0 +1,2423 @@
+                let farmed = self
+                    .farm_token_rates
+                    .iter()
+                    .map(|rate| U128::from(farmed_tokens(v.farmed, *rate)))
+                    .collect();
+                return Some( P4Status {
+                    stake_tokens: v.staked,
+                    stake: v.min_stake.into(),
+                    farmed_units: v.farmed.into(),
+                    farmed_tokens: farmed,
+                    boost_nfts: v.boost_nft,
+                    timestamp: self.farming_start + r0 * ROUND,
+                    total_cheddar_staked: v.total_cheddar_staked.into()
+                });
+            }
+            None => None,
+        };
+    }
+
+    // ******************* //
+    // transaction methods //
+    // ******************* //
+
+    /// withdraw NFT to a destination account using the `nft_transfer` method.
+    /// This function is considered safe and will work when contract is paused to allow user
+    /// to withdraw his NFTs.
+    #[payable]
+    pub fn withdraw_boost_nft(&mut self) {
+        assert_one_yocto();
+        let user = env::predecessor_account_id();
+        let mut vault = self.get_vault(&user);
+        self._withdraw_boost_nft(&user, &mut vault);
+    }
+
+    /// Deposit native near during the setup phase for farming rewards.
+    /// Panics when the deposit was already done or the setup is completed.
+    #[payable]
+    pub fn setup_deposit_near(&mut self) {
+        self._setup_deposit(&near(), env::attached_deposit())
+    }
+    /// FT Receiver `setup deposit` scenario
+    /// Panics on failed `unwrap()` if FT not set in `Contract.farm_tokens`
+    pub(crate) fn _setup_deposit(&mut self, token: &AccountId, amount: u128) {
+        assert!(
+            !self.setup_finalized,
+            "setup deposits must be done when contract setup is not finalized"
+        );
+        let token_i = find_acc_idx(token, &self.farm_tokens);
+        let total_rounds = round_number(self.farming_start, self.farming_end, self.farming_end);
+        let expected = farmed_tokens(
+            u128::from(total_rounds) * self.farm_unit_emission,
+            self.farm_token_rates[token_i],
+        );
+        assert_eq!(
+            self.farm_deposits[token_i], 0,
+            "deposit already done for the given token"
+        );
+        assert_eq!(
+            amount, expected,
+            "Expected deposit for token {} is {}, got {}",
+            self.farm_tokens[token_i], expected, amount
+        );
+        self.farm_deposits[token_i] = amount;
+    }
+
+    /// FT Receiver `cheddar stake` scenario
+    pub(crate) fn setup_cheddar_payment(&mut self, sender_id: &AccountId, amount: u128) {
+        self.assert_is_active();
+        let user = sender_id.clone();
+        let mut v = self.get_vault(&user);
+
+        // Expected cheddar for stake per one token
+        let expected = self.cheddar_rate;
+        assert_eq!(
+            expected, amount, 
```

Comment:
```suggestion
        assert_true!(
            expected >= amount, 
```

Reply:
Ok, I will update. BTW, `assert!()` doing the same as `assert_true!()` isn’t it?

Reply:
Up

Reply:
re `assert!()` - I think yes.